### DISCUSSION TITLE
Character reverts to DefaultObject

### DISCUSSION THREAD

**Discussion by kovitikus** [evennia/evennia#2380] [Evennia Questions] [answered]
I'm posting this here first, because I'm not sure if this is an issue that can be or should be addressed, but it is definitely a bug, possibly an error entirely the user's fault.

#### Describe the bug
All I did was change some strings in my exits, adding a period to the list of exits generated and sent back to the room for return_appearance. Yet for some reason, when I reloaded, my character was suddenly the wrong typeclass. There's absolutely no reason for this to have happened. The typeclass path on my character object remains `characters.characters.Character` and that path is set in my server settings as the default as well.

Actually, it seems I left a rogue period inside a spot it shouldn't be, an fstring. Somehow this fstring syntax error lead to the character object reverting back to DefaultObject on reload, which I'm assuming has to do with fstrings being evaluated at runtime.
![rogue period inside fstring](https://i.imgur.com/XJAjgLm.png)
![rogue period traceback](https://i.imgur.com/7dMI9kg.png)
This traceback was the result of attempting to swap the typeclass back to `characters.characters.Character`.

```
# Typeclass for character objects linked to an account (fallback)
BASE_CHARACTER_TYPECLASS = "characters.characters.Character"
```

#### To Reproduce
**No clue.**

#### Environment, Evennia version, OS etc
         Evennia version: 0.9.5 (rev a07def36a)
         OS: nt
         Python: 3.7.4
         Twisted: 20.3.0
         Django: 2.2.16

#### Additional context
There was another day where this seems to have happened to someone new to Evennia.

https://irclogs.jackgrigg.com/irc.freenode.net/evennia/2021-03-27#i_4706337

Here's my activity at the time it happened.
```
 Server restart initiated ...
 ... Server restarted.
Traceback (most recent call last):
  File "d:\muddev\evennia\evennia\commands\cmdhandler.py", line 634, in _run_command
    abort = yield cmd.at_pre_cmd()
  File ".\commands\command.py", line 44, in at_pre_cmd
    self.caller.status.afk_check()
AttributeError: 'DefaultObject' object has no attribute 'status'

An untrapped error occurred.
(Traceback was logged 21-04-11 15:31:45-04).
Traceback (most recent call last):
  File "d:\muddev\evennia\evennia\commands\cmdhandler.py", line 634, in _run_command
    abort = yield cmd.at_pre_cmd()
  File ".\commands\command.py", line 44, in at_pre_cmd
    self.caller.status.afk_check()
AttributeError: 'DefaultObject' object has no attribute 'status'

An untrapped error occurred.
(Traceback was logged 21-04-11 15:31:53-04).
--------------------------------------------------------------------------------------------
Name/key: Kovitikus (#1)
Typeclass: DefaultObject (characters.characters.Character)
Session id(s): #1
Account: Kovitikus (#1)
  Account Typeclass: Account (typeclasses.accounts.Account)
  Account Permissions: <Superuser>
Location: a sacred grove of ancient wood (#330)
Home: Main Office (#4)
Permissions: player <Superuser>
Locks: 
  boot:false(); call:false(); control:perm(Developer); delete:false();
  drop:holds(); edit:false(); examine:perm(Developer); get:false(); msg:all();
  puppet:false(); tell:perm(Admin); view:all()
Stored Cmdset(s): 
  commands.default_cmdsets.CharacterCmdSet [DefaultCharacter] (Union, prio 0)
  travel.travel_cmdset.TravelCmdSet [card_dir] (Union, prio 0)
Merged Cmdset(s): 
  <Current merged cmdset> (Union prio 101)
  evennia.commands.cmdset.CmdSet [ExitCmdSet] (Union prio 101, duplicates: T)
  evennia.commands.cmdset.CmdSet [ChannelCmdSet] (Union prio 101, duplicates: T)
  commands.default_cmdsets.CharacterCmdSet [DefaultCharacter] (Union prio 0)
  travel.travel_cmdset.TravelCmdSet [card_dir] (Union prio 0)
  commands.default_cmdsets.AccountCmdSet [DefaultAccount] (Union prio -10)
  commands.default_cmdsets.SessionCmdSet [DefaultSession] (Union prio -20)
Commands available to Kovitikus (result of Merged CmdSets): 
  @grant-sp, a sparsely-populated fledgling forest, a sparsely-populated
  fledgling forest, abandonfailedtraveller, about, access, accounts, addcom,
  afktimer, alias, allcom, approach, ban, batchcode, batchcommands, boot, buy,
  cboot, ccreate, cdesc, cdestroy, cemit, channels, charcreate, chardelete,
  chargen, clock, cmdsets, color, convertcoin, copy, cpattr, create, cwho,
  delcom, desc, destroy, dig, drop, e, emit, equip, examine, find, force, get,
  give, grapevine2chan, group, heal, help, home, ic, inhand, instance,
  inventory, irc2chan, ircstatus, kneel, learn, lie, light, link, lock, look,
  mudinfo, mvattr, n, name, ne, nick, nw, objects, ooc, open, option, page,
  password, perm, pose, public, put, py, quell, quit, reload, reset, retreat,
  rss2chan, s, say, script, scripts, se, server, service, sessions, set,
  setdesc, sethelp, sethome, shutdown, sit, skills, spawn, stand, stock, stow,
  style, sw, swat, tag, tel, testy, tickers, time, tunnel, typeclass, unban,
  ungroup, unlink, unwield, userpassword, w, wall, whisper, who, wield, wipe
Persistent attribute(s): 
  afk = False
  afk_timer = 600
  approached = []
  busy = False
  coin = {'plat': 0, 'gold': 0, 'silver': 0, 'copper': 0}
  desc = This is User #1.
  equipment = {'head': None, 'neck': None, 'shoulders': None, 'chest': None, 'arms': No[...]
  feinted = None
  followers = []
  hands = {'main': None, 'off': None}
  hands_desc = {'main': 'right', 'off': 'left'}
  hp = {'max_hp': 100, 'current_hp': 100}
  instances = {'1_2021-04-03 08:28:17.674073': {'creator': 'Kovitikus', 'epoch_creation[...]
  inventory_slots = {'max_slots': 50, 'occupied_slots': 0}
  kneeling = False
  ko = False
  leader = None
  lying = False
  prelogout_location = Main Office
  sitting = False
  stance = None
  standing = True
  wielding = {'main': None, 'off': None, 'both': None}
Contents: a tattered green bag(#3)
--------------------------------------------------------------------------------------------
```

**Comment by kovitikus** [comment]
I reloaded and the typeclass has reverted back to Character. I'm not sure if that's because the typeclass command I attempted to enter actually went through before the reload, or if fixing the fstring somehow fixed the character's typeclass.

```
--------------------------------------------------------------------------------------------
Name/key: Kovitikus (#1)
Typeclass: Character (characters.characters.Character)
Session id(s): #1
Account: Kovitikus (#1)
  Account Typeclass: Account (typeclasses.accounts.Account)
  Account Permissions: <Superuser>
Location: a sacred grove of ancient wood (#330)
Home: Main Office (#4)
Permissions: player <Superuser>
Locks: 
  boot:false(); call:false(); control:perm(Developer); delete:false();
  drop:holds(); edit:false(); examine:perm(Developer); get:false(); msg:all();
  puppet:false(); tell:perm(Admin); view:all()
Stored Cmdset(s): 
  commands.default_cmdsets.CharacterCmdSet [DefaultCharacter] (Union, prio 0)
  travel.travel_cmdset.TravelCmdSet [card_dir] (Union, prio 0)
Merged Cmdset(s): 
  <Current merged cmdset> (Union prio 101)
  evennia.commands.cmdset.CmdSet [ExitCmdSet] (Union prio 101, duplicates: T)
  evennia.commands.cmdset.CmdSet [ChannelCmdSet] (Union prio 101, duplicates: T)
  commands.default_cmdsets.CharacterCmdSet [DefaultCharacter] (Union prio 0)
  travel.travel_cmdset.TravelCmdSet [card_dir] (Union prio 0)
  commands.default_cmdsets.AccountCmdSet [DefaultAccount] (Union prio -10)
  commands.default_cmdsets.SessionCmdSet [DefaultSession] (Union prio -20)
Commands available to Kovitikus (result of Merged CmdSets): 
  @grant-sp, a sparsely-populated fledgling forest, a sparsely-populated
  fledgling forest, abandonfailedtraveller, about, access, accounts, addcom,
  afktimer, alias, allcom, approach, ban, batchcode, batchcommands, boot, buy,
  cboot, ccreate, cdesc, cdestroy, cemit, channels, charcreate, chardelete,
  chargen, clock, cmdsets, color, convertcoin, copy, cpattr, create, cwho,
  delcom, desc, destroy, dig, drop, e, emit, equip, examine, find, force, get,
  give, grapevine2chan, group, heal, help, home, ic, inhand, instance,
  inventory, irc2chan, ircstatus, kneel, learn, lie, light, link, lock, look,
  mudinfo, mvattr, n, name, ne, nick, nw, objects, ooc, open, option, page,
  password, perm, pose, public, put, py, quell, quit, reload, reset, retreat,
  rss2chan, s, say, script, scripts, se, server, service, sessions, set,
  setdesc, sethelp, sethome, shutdown, sit, skills, spawn, stand, stock, stow,
  style, sw, swat, tag, tel, testy, tickers, time, tunnel, typeclass, unban,
  ungroup, unlink, unwield, userpassword, w, wall, whisper, who, wield, wipe
Persistent attribute(s): 
  afk = False
  afk_timer = 600
  approached = []
  busy = False
  coin = {'plat': 0, 'gold': 0, 'silver': 0, 'copper': 0}
  desc = This is User #1.
  equipment = {'head': None, 'neck': None, 'shoulders': None, 'chest': None, 'arms': No[...]
  feinted = None
  followers = []
  hands = {'main': None, 'off': None}
  hands_desc = {'main': 'right', 'off': 'left'}
  hp = {'max_hp': 100, 'current_hp': 100}
  instances = {'1_2021-04-03 08:28:17.674073': {'creator': 'Kovitikus', 'epoch_creation[...]
  inventory_slots = {'max_slots': 50, 'occupied_slots': 0}
  kneeling = False
  ko = False
  leader = None
  lying = False
  prelogout_location = Main Office
  sitting = False
  stance = None
  standing = True
  wielding = {'main': None, 'off': None, 'both': None}
Contents: a tattered green bag(#3)
--------------------------------------------------------------------------------------------
```

**Reply by RandomEC** [reply]
Kovitikus, how did you get everything to try to reload again? I've had this same problem. I've tried stripping the character class back to its roots to try to figure out what the problem was, but neither reload nor shutting it down and restarting seems to let the character class try to reload again.

**Reply by Tegiminis** [reply]
You might have a circular import. That's usually how I get my "revert to DefaultObject" problems

**Reply by kovitikus** [reply]
> Kovitikus, how did you get everything to try to reload again? I've had this same problem. I've tried stripping the character class back to its roots to try to figure out what the problem was, but neither reload nor shutting it down and restarting seems to let the character class try to reload again.

I removed the rogue period I had misplaced. I meant to put a period at the end of the string, but I instead put it inside of the fstring variable section.

**Reply by RandomEC** [reply]
I figured it out. I thought that fixing the problem in the Character class would cause the game to revert on reloading to my character BEING a Character. I did not even realize that there was a command for setting typeclass. I found that later last night and tried it, and got an error message telling me what was still wrong. 

Edit: in case it is useful to anyone else, what I did here was, faced with the error that my character was reading as a DefaultObject, I tried "update"-ing it, which was the exact wrong thing to do. Once I did that, it fully reset its typeclass to DefaultObject so that even when I reloaded, it came back as that. I was able to fix this goof by, as noted above, simply using the typeclass command after fixing the error in the Character class.

Thanks for the attempted help!

Random

**Comment by Griatch** [comment, marked as answer]
This is not a bug. What happens is that your misplaced comma makes the module an invalid Python module - that is, Evennia can't even open it since it immediately errors on on import. For such an important typeclass, this would completely bring down the server (and depending on timing and where in the module the error is, this may still be unavoidable), so instead Evennia tries very hard to offer a solution - it logs the error and falls back to the one thing it can expect to work - the DefaultObject parent class, which at least allows the server to come back up - hopefully with less effect on other users, depending on the typeclass affected (remember this can happen for any of your typeclasses, the Character is just the most immediately visual case). 

This is just a temporary change, so once you fix the critical error and reload, your actual typeclass (`Character` in this case) will be able to load normally and all will be well. This behavior is exemplified e.g. in the beginning tutorial [here](https://www.evennia.com/docs/latest/First-Steps-Coding.html#troubleshooting-updating-yourself). 

Whereas fixing the logged error and reloading should be the normal process for handling any sudden evennia error like this, the typeclass' temporary recovery fallback should probably be more explicitly explained in the Typeclass documentation as well. It's hard to add more log info to the log for all cases because this type of error interrupts Python quite rudely and can come from a wide variety of sources.

**Reply by kovitikus** [reply]
Interesting how the entire Character typeclassed failed due to an error in my TravelHandler. But I suppose that a property is a defining part of the typeclass.

Luckily I tried to do the typeclass command and it TOLD me where the error was occurring. Who knows how long I would have been confused otherwise.

Thanks for the info. 👍🏼

**Reply by Griatch** [reply]
The typeclass command attempts to re-import the typeclass from file and reports its failure to do so. You should have seen the error also in the main server log when you reloaded (this is when the class first fell back to the DefaultObject parent).

**Reply by kovitikus** [reply]
Yeah, it's definitely in the log file. :)
